akka-projection-core/src/main/scala/akka/projection/scaladsl/StatefulHandler.scala

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.projection.scaladsl
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.util.Failure
+import scala.util.Success
+
+import akka.Done
+import akka.annotation.ApiMayChange
+
+@ApiMayChange abstract class StatefulHandler[State, Envelope](implicit ec: ExecutionContext) extends Handler[Envelope] {
+
+  private var state: Future[State] = _
+
+  /**
+   * Invoked to load the initial state when the projection is started or if previous `process` failed.
+   */
+  def initialState(): Future[State]
+
+  /**
+   * The `process` method is invoked for each `Envelope`.
+   * One envelope is processed at a time. The returned `Future[State]` is to be completed when the processing
+   * of the `envelope` has finished. It will not be invoked with the next envelope until after the returned
+   * `Future` has been completed.
+   *
+   * The `state` is the completed value of the previously returned `Future[State]` or the `initialState`.
+   * If the previously returned `Future[State]` failed it will call `initialState`
+   * again and use that value.
+   */
+  def process(state: State, envelope: Envelope): Future[State]
+
+  /**
+   * Calls [[StatefulHandler.initialState]] when the projection is started.
+   */
+  final override def start(): Future[Done] = {
+    state = initialState()
+    state.map(_ => Done)
+  }
+
+  /**
+   * Calls `process(state, envelope)` with the completed value of the previously returned `Future[State]`
+   * or the `initialState`. If the previously returned `Future[State]` failed it will call `initialState`
+   * again and use that value.
+   */
+  final override def process(envelope: Envelope): Future[Done] = {
+    val newState = state.value match {
```

Comment:
Why `state.value` and not `flatMap/recoverWith`?

Reply:
because we know that it's always completed here

Reply:
Ok, I see. And then you don't pay the flatMap tax?